include/trial/protocol/json/partial/skip.hpp

```diff
@@ -0,0 +1,72 @@
+#ifndef TRIAL_PROTOCOL_JSON_PARTIAL_SKIP_HPP
+#define TRIAL_PROTOCOL_JSON_PARTIAL_SKIP_HPP
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Copyright (C) 2018 Vinícius dos Santos Oliveira <vini.ipsmaker@gmail.com>
+//
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          http://www.boost.org/LICENSE_1_0.txt)
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include <trial/protocol/json/reader.hpp>
+
+namespace trial
+{
+namespace protocol
+{
+namespace json
+{
+namespace partial
+{
+
+template<class CharT>
+typename basic_reader<CharT>::view_type skip(basic_reader<CharT> &reader)
```

Comment:
I suggest that we add a `std::error_code&` output parameter which is set when `reader.symbol() == token::symbol::error` or if skipping fails because the user has fed an invalid or incomplete JSON buffer into `reader` (e.g. `"["`.)

We can also add a throwing version of the function.

Reply:
Good point.

> I suggest that we add a `std::error_code&` output parameter which is set when `reader.symbol() == token::symbol::error`

You raised a good point, but the solution isn't easy. I don't feel comfortable with any of the options.

Conceptually, the `skip` has done what it was told to do. It has succeeded. So, if it succeeded, why would it set an error flag? The user would have to check `reader.symbol()` after the call _anyway_ to continue consuming the document.

But I understand the reasoning to set an error flag _even_ if succeeded, but current token is `error`. User could forget to check `reader.symbol()` after the token (we can argue against this because the user just _has to_ check on the current token before processing it, but [this “forget to check” could happen on the sample code I just gave earlier and it would trigger a bug, yes](https://github.com/breese/trial.protocol/issues/23#issuecomment-365745929)).

I think I'll agree with you and go ahead on setting an error, but I need at least a few days to think more about the problem. The rest of the review is easier to resolve.

> or if skipping fails because the user has fed an invalid or incomplete JSON buffer into `reader` (e.g. `"["`.)

Okay, it makes sense here. The user can't know currently if `skip` succeeded or failed when `reader.symbol() == error` after the call.